File: LeetCode/Study Plan/Practice Problems/Medium/Rotate Image/Go/rotate_image.go
package main

import "fmt"

func rotate_image(arr [][]int) {
	//transpose + reflex = will rotate the matrix by 90
	//transpose
	n, m := len(arr), len(arr[0])
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if i < j {
				arr[i][j], arr[j][i] = arr[j][i], arr[i][j]
			}
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < m/2; j++ {
			arr[i][j], arr[i][n-j-1] = arr[i][n-j-1], arr[i][j]
		}
	}
	fmt.Println(arr)
}

func main() {
	rotate_image([][]int{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}})
}

File: LeetCode/Study Plan/Practice Problems/Medium/Rotate Image/Java/rotateImage.java
class Solution {
    public void rotate(int[][] matrix) {
        //transpose + reflex gives rotation by 90 degrees.
        int n = matrix.length;
        
        //transpose
        for (int i=0; i<n; i++) {
            for (int j=0; j<n; j++) {
                if (i<j) {
                    int temp = matrix[i][j];
                    matrix[i][j] = matrix[j][i];
                    matrix[j][i] = temp; 
                }
            }
        }

        //reflex
        for (int i=0; i<n; i++) {
            for (int j=0; j<n/2; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[i][n-j-1];
                matrix[i][n-j-1] = temp;
            }
        }
    }
}
